global_ens: add restart capability to atmos grid2obs stats jobs

GwenChen-NOAA opened the pull request:
## Description of Changes
This PR adds restart capability to atmos grid2obs stats jobs for the global_ens component. This PR addresses Issue #532. 

## Developer Questions and Checklist
* Is this a high priority PR? If so, why and is there a date it needs to be merged by? Yes.
* Do you have any planned upcoming annual leave/PTO? No.
* Are there any changes needed in the times when the jobs are supposed to run/kick-off? No.
  
- [X] The code changes follow [NCO's EE2 Standards](https://www.nco.ncep.noaa.gov/idsb/implementation_standards/ImplementationStandards.v11.0.0.pdf).
- [X] Developer's name is removed throughout the code and have used `${USER}` where necessary throughout the code.
- [X] References the feature branch for `HOMEevs` are removed from the code.
- [X] J-Job environment variables, COMIN and COMOUT directories, and output follow what has been [defined](https://docs.google.com/document/d/1JWg_4q80aYmmAoD21GFjp9R9y5-3w7WGM3-0HJk0Pjs/edit#heading=h.7ysbr191vzu4) for EVS.
- [X] Jobs over 15 minutes in runtime have restart capability.
- [X] If applicable, changes in the `dev/drivers/scripts` or `dev/modulefiles` have been made in the corresponding `ecf/scripts` and `ecf/defs/evs-nco.def`? 
- [X] Jobs contain the appropriate file checking and don't run METplus for any missing data.
- [X] Code is using METplus wrappers structure and not calling MET executables directly.
- [X] Log is free of any ERRORs or WARNINGs.

## Testing Instructions
(1) Set up jobs
a. Symlink the EVS_fix directory locally as "fix".
b. In the driver scripts, edit the following environment variables:

HOMEevs - set to your test EVS directory
COMIN - set to /lfs/h2/emc/vpppg/noscrub/emc.vpppg/${NET}/$evs_ver_2d
COMOUT - set to your test output directory

(2) Run jobs
Run the following jobs in dev/drivers/scripts/stats/global_ens for any VDATE:

qsub jevs_global_ens_cmce_atmos_grid2obs_stats.sh
qsub jevs_global_ens_ecme_atmos_grid2obs_stats.sh
qsub jevs_global_ens_gefs_atmos_grid2obs_stats.sh
qsub jevs_global_ens_naefs_atmos_grid2obs_stats.sh

Log files should be checked for free of errors.

(3) Test restart capability
After a successful full run, save the final stat file for comparison. Run the job again and randomly stop the job using qdel. Then, resubmit the job using qsub. The resubmitted job should run shorter than the full run, and the final stat file from the resubmitted job should be the same from the full run. The jobs run from about 1 hour (for naefs) to 2+ hours (for gefs).

GwenChen-NOAA committed:
Add restart for grid2obs stats jobs

AliciaBentley-NOAA commented:
@GwenChen-NOAA I hope to get started on this PR today! Can you please sync your fork? Thanks!

GwenChen-NOAA committed:
Merge branch 'NOAA-EMC:develop' into grid2obs_restart

GwenChen-NOAA commented:
> @GwenChen-NOAA I hope to get started on this PR today! Can you please sync your fork? Thanks!

Done.

AliciaBentley-NOAA commented:
@GwenChen-NOAA I have started to test this PR and submitted the following jobs. Please check the logfile and output when they have completed (all are about 5 minutes in). I also turned KEEPDATA on for the purposes of troubleshooting during testing.  

### jevs_global_ens_cmce_atmos_grid2obs_stats.sh
**Log file:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/
**Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
**DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/jevs_global_ens_cmce_atmos_grid2obs_stats.28337269.dbqs01

### jevs_global_ens_ecme_atmos_grid2obs_stats.sh
**Log file**: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/
**Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
**DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/jevs_global_ens_ecme_atmos_grid2obs_stats.28337270.dbqs01

### jevs_global_ens_gefs_atmos_grid2obs_stats.sh
**Log file:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/
**Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
**DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/jevs_global_ens_gefs_atmos_grid2obs_stats.28337271.dbqs01

### jevs_global_ens_naefs_atmos_grid2obs_stats.sh
**Log file:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/
**Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
**DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/jevs_global_ens_naefs_atmos_grid2obs_stats.28337272.dbqs01

BinbinZhou-NOAA commented:
@GwenChen-NOAA  I just looked at your working directories, and noticed that the size of those completed files are 0.
In the testing of my narre PR, today,   I got the same issue and Perry just told me that this will not be accepted by NCO. 
Perry suggested just adding one line of text to the completed files, like ".... is completed"

GwenChen-NOAA committed:
Add text to the completed files

GwenChen-NOAA committed:
Merge branch 'grid2obs_restart' of https://github.com/GwenChen-NOAA/EVS into grid2obs_restart

GwenChen-NOAA commented:
> @GwenChen-NOAA I just looked at your working directories, and noticed that the size of those completed files are 0. In the testing of my narre PR, today, I got the same issue and Perry just told me that this will not be accepted by NCO. Perry suggested just adding one line of text to the completed files, like ".... is completed"

@BinbinZhou-NOAA, thanks for bringing this up! I added a line of text to the completed files.

@AliciaBentley-NOAA, I also updated the code for the grid2grid and precip jobs. If you want to retest those jobs, run the following jobs in dev/drivers/scripts/stats/global_ens:

qsub jevs_global_ens_cmce_atmos_grid2grid_stats.sh
qsub jevs_global_ens_ecme_atmos_grid2grid_stats.sh
qsub jevs_global_ens_gefs_atmos_grid2grid_stats.sh
qsub jevs_global_ens_naefs_atmos_grid2grid_stats.sh
qsub jevs_global_ens_cmce_atmos_precip_stats.sh
qsub jevs_global_ens_ecme_atmos_precip_stats.sh
qsub jevs_global_ens_gefs_atmos_precip_stats.sh
qsub jevs_global_ens_naefs_atmos_precip_stats.sh

AliciaBentley-NOAA commented:
@GwenChen-NOAA Sure! Should I clean up the first PR test (i.e., log files, output, DATA) and start the full test over with your updated code?

GwenChen-NOAA commented:
> @GwenChen-NOAA Sure! Should I clean up the first PR test (i.e., log files, output, DATA) and start the full test over with your updated code?

Yes, if you want to restart the test, it needs to start from fresh. Since the grid2obs jobs run quite long, it may not finish today. You may start again tomorrow morning, then no need to clean up.

AliciaBentley-NOAA commented:
Thanks, @GwenChen-NOAA. I'll kick off the clean test tomorrow morning! Do you think that there will be any problem with the VDATE if I start jobs between 7:30am and 8:00am?

GwenChen-NOAA commented:
> Thanks, @GwenChen-NOAA. I'll kick off the clean test tomorrow morning! Do you think that there will be any problem with the VDATE if I start jobs between 7:30am and 8:00am?

I don't think so. I run my tests in cron at mid-night.

AliciaBentley-NOAA commented:
@GwenChen-NOAA I git pulled in your updated code. All jobs original listed to test in this PR, as well as all the jobs in your previous comment (atmos_grid2grid and precip), have been kicked off. Please check the logfile and output when they have completed. I also turned on KEEPDATA for the purposes of troubleshooting during testing.

### jevs_global_ens_cmce_atmos_grid2obs_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_cmce_atmos_grid2obs_stats.o28461526
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

### jevs_global_ens_ecme_atmos_grid2obs_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_ecme_atmos_grid2obs_stats.o28461528
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

### jevs_global_ens_gefs_atmos_grid2obs_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_gefs_atmos_grid2obs_stats.o28461531 
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

### jevs_global_ens_naefs_atmos_grid2obs_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_naefs_atmos_grid2obs_stats.o28461535
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

---------------------------------------------------------

### jevs_global_ens_cmce_atmos_grid2grid_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_cmce_atmos_grid2grid_stats.o28461559
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

### jevs_global_ens_ecme_atmos_grid2grid_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_ecme_atmos_grid2grid_stats.o28461561
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

### jevs_global_ens_gefs_atmos_grid2grid_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_gefs_atmos_grid2grid_stats.o28461566 
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

### jevs_global_ens_naefs_atmos_grid2grid_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_naefs_atmos_grid2grid_stats.o28461568
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

---------------------------------------------------------

### jevs_global_ens_cmce_atmos_precip_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_cmce_atmos_precip_stats.o28461675
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

### jevs_global_ens_ecme_atmos_precip_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_ecme_atmos_precip_stats.o28461676
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

### jevs_global_ens_gefs_atmos_precip_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_gefs_atmos_precip_stats.o28461684 
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

### jevs_global_ens_naefs_atmos_precip_stats.sh
Log file: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_naefs_atmos_precip_stats.o28461687
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

GwenChen-NOAA commented:
@AliciaBentley-NOAA, all jobs finished successfully without any errors in the log files. The output stat files look good to me too. Please save the output directories under /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens (I usually just `mv atmos.20250401 atmos.20250401.saved` for example ). You may start the restart test. Thanks!

AliciaBentley-NOAA commented:
Thanks, @GwenChen-NOAA ! I can confirm that I do not see any issues during my search of the log files. I have also moved the stats directories as you suggested to keep them. Can you please remind me of two things in order to test restart?

1) Which jobs should I be running again? All of them, or just the jobs originally intended for this PR?

2) How long should I let them run before stopping the jobs?

GwenChen-NOAA commented:
@AliciaBentley-NOAA, you can test all jobs if you'd like, but definitely the grid2obs jobs. You can kill the jobs anytime you want. It's easier to see the benefits of restart if you let it run longer. I usually let the jobs run about half way. Below are the estimated job runtimes:

jevs_global_ens_cmce_atmos_grid2grid_stats.sh	0:28
jevs_global_ens_cmce_atmos_grid2obs_stats.sh	1:20
jevs_global_ens_cmce_atmos_precip_stats.sh	        0:04
jevs_global_ens_ecme_atmos_grid2grid_stats.sh	1:13
jevs_global_ens_ecme_atmos_grid2obs_stats.sh	1:53
jevs_global_ens_ecme_atmos_precip_stats.sh	        0:10
jevs_global_ens_gefs_atmos_grid2grid_stats.sh	0:55
jevs_global_ens_gefs_atmos_grid2obs_stats.sh	2:06
jevs_global_ens_gefs_atmos_precip_stats.sh	        0:06
jevs_global_ens_naefs_atmos_grid2grid_stats.sh	0:08
jevs_global_ens_naefs_atmos_grid2obs_stats.sh	0:45
jevs_global_ens_naefs_atmos_precip_stats.sh	        0:04

AliciaBentley-NOAA commented:
@GwenChen-NOAA I have kicked off the jobs you listed above with VDATE=20250401 and wallclock times that are about half of what the full runtime is. Many jobs are still running. 

A few things to note. The following jobs have runtime less than 15 minutes and do **not** need restart capabilities. That said, I've still tested them like you suggested, so we should examine the output. 

```
jevs_global_ens_cmce_atmos_precip_stats.sh 0:04
jevs_global_ens_ecme_atmos_precip_stats.sh 0:10
jevs_global_ens_gefs_atmos_precip_stats.sh 0:06
jevs_global_ens_naefs_atmos_grid2grid_stats.sh 0:08
jevs_global_ens_naefs_atmos_precip_stats.sh 0:04
```

I also noticed that the wallclock times alloted are kind of all over the place. Some request much more than the runtime, while others are cutting it really close. If you could please update the following dev driver wallclock times in this PR (I think these ones are the worst offenders), that would be great. Please see the EVS Best Practices document for information from NCO about ideal wallclock times. 

```
jevs_global_ens_cmce_atmos_grid2grid_stats.sh (runtime 0:28, requests 1:30, decrease wallclock requested)
jevs_global_ens_cmce_atmos_precip_stats.sh (runtime 0:04, requests 0:30, decrease wallclock requested)
jevs_global_ens_ecme_atmos_grid2obs_stats.sh (runtime 1:53, requests 2:00, **increase** wallclock requested)
jevs_global_ens_gefs_atmos_precip_stats.sh (runtime 0:06, requests 0:30, decrease wallclock requested)
jevs_global_ens_naefs_atmos_grid2grid_stats.sh (runtime 0:08, requests 0:30, decrease wallclock requested)
jevs_global_ens_naefs_atmos_precip_stats.sh (runtime 0:04, requests 0:30, decrease wallclock requested)
```

-------------------------------------------------------------------------------------

Jobs started:

```
jevs_global_ens_cmce_atmos_grid2grid_stats.sh
jevs_global_ens_cmce_atmos_grid2obs_stats.sh
jevs_global_ens_cmce_atmos_precip_stats.sh
jevs_global_ens_ecme_atmos_grid2grid_stats.sh
jevs_global_ens_ecme_atmos_grid2obs_stats.sh
jevs_global_ens_ecme_atmos_precip_stats.sh
jevs_global_ens_gefs_atmos_grid2grid_stats.sh
jevs_global_ens_gefs_atmos_grid2obs_stats.sh
jevs_global_ens_gefs_atmos_precip_stats.sh
jevs_global_ens_naefs_atmos_grid2grid_stats.sh
jevs_global_ens_naefs_atmos_grid2obs_stats.sh
jevs_global_ens_naefs_atmos_precip_stats.sh
```

(use `ls -lath` to see log files from today near the top)
Log files: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/
Output: /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens/
DATA: /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

AliciaBentley-NOAA commented:
@GwenChen-NOAA No need to look at the output just yet! I didn't get pull in your updates first. I'll let you know when I kick them off again. Sorry about that!

AliciaBentley-NOAA commented:
Well, before I qdel anything, let me check. I may have done the git pull yesterday... two seconds...

AliciaBentley-NOAA commented:
@GwenChen-NOAA Good news! I pulled the changes in 2 days ago, I just still had written to do it in my notes by accident! Sorry for the false alarm!

One quick question: Is there a third "precip" ush/ file that needs to be updated? I see a grid2grid and a grid2obs in your commit to make sure the files aren't 0-sized, so I'm just wondering if there is a corresponding precip ush/ script that needs to be updated too. Please let me know! Thanks!

CC @PerryShafran-NOAA

PerryShafran-NOAA commented on ush/global_ens/evs_global_ens_atmos_grid2grid.sh:
```diff
@@ -511,6 +513,7 @@ if [ $verify = precip ] ; then
         # Indicate sub-task is completed for restart
 	echo ">$WORK/run_${modnam}_ccpa${apcp}_valid_at_t${vhour}z_fhr${lead_hr}_${metplus_job}.completed" >> run_${modnam}_ccpa${apcp}_valid_at_t${vhour}z_${metplus_job}.sh
+        echo "echo "${modnam}_ccpa${apcp}_valid_at_t${vhour}z_fhr${lead_hr}_${metplus_job} task is completed" >> $WORK/run_${modnam}_ccpa${apcp}_valid_at_t${vhour}z_fhr${lead_hr}_${metplus_job}.completed" >> run_${modnam}_ccpa${apcp}_valid_at_t${vhour}z_${metplus_job}.sh
 
```
@AliciaBentley-NOAA @GwenChen-NOAA Alicia had asked about changing a third ush script for precip.  I think that the change here on line 516 relates to precip.  I think that all the precip jobs use the evs_global_ens_atmos_grid2grid.sh for both g2g and for precip. I think we are good on this aspect.

GwenChen-NOAA replied:
Thanks, @PerryShafran-NOAA! You got it. Precip jobs are included in the grid2grid ush script. This is why restart is added to precip jobs although it is not needed, to keep the code consistent.

GwenChen-NOAA commented:
> I also noticed that the wallclock times alloted are kind of all over the place. Some request much more than the runtime, while others are cutting it really close. If you could please update the following dev driver wallclock times in this PR (I think these ones are the worst offenders), that would be great. Please see the EVS Best Practices document for information from NCO about ideal wallclock times.
> 
> ```
> jevs_global_ens_cmce_atmos_grid2grid_stats.sh (runtime 0:28, requests 1:30, decrease wallclock requested)
> jevs_global_ens_cmce_atmos_precip_stats.sh (runtime 0:04, requests 0:30, decrease wallclock requested)
> jevs_global_ens_ecme_atmos_grid2obs_stats.sh (runtime 1:53, requests 2:00, **increase** wallclock requested)
> jevs_global_ens_gefs_atmos_precip_stats.sh (runtime 0:06, requests 0:30, decrease wallclock requested)
> jevs_global_ens_naefs_atmos_grid2grid_stats.sh (runtime 0:06, requests 0:30, decrease wallclock requested)
> jevs_global_ens_naefs_atmos_precip_stats.sh (runtime 0:04, requests 0:30, decrease wallclock requested)
> ```

@AliciaBentley-NOAA, these runtime estimates are without restart. Adding restart capability usually increases runtime and sometime memory usage. I will wait after restart tested successfully to adjust the numbers. Is there a minimum wallclock time setting?

AliciaBentley-NOAA commented:
@GwenChen-NOAA In the [EVS Best Practices document](https://docs.google.com/document/d/1Tve8cxESi-uaYcELHWk2Z-imNfqpHq3i2LKyRkZuEEY/edit?tab=t.0#heading=h.ey3fecfgzjs7), you'll see that the lowest wallclock time allowed is 10 minutes. A 4-minute jobs shouldn't request 30 minutes of wallclock time according to NCO. You can make adjustments to the dev driver wallclock times after the restart capability is tested, or you can make them now and I just won't git pull them until the restart is tested. 

All jobs were successfully interrupted part way through. How do the restart files look to you?

(use ls -lath to see log files from today near the top)
**Log files:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/
**Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens/
**DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

AliciaBentley-NOAA commented:
@GwenChen-NOAA In addition to letting me know when I can run the remainder of the test, please sync your fork. A couple of PRs were merged this morning. Thanks!

GwenChen-NOAA commented:
> @GwenChen-NOAA In the [EVS Best Practices document](https://docs.google.com/document/d/1Tve8cxESi-uaYcELHWk2Z-imNfqpHq3i2LKyRkZuEEY/edit?tab=t.0#heading=h.ey3fecfgzjs7), you'll see that the lowest wallclock time allowed is 10 minutes. A 4-minute jobs shouldn't request 30 minutes of wallclock time according to NCO. You can make adjustments to the dev driver wallclock times after the restart capability is tested, or you can make them now and I just won't git pull them until the restart is tested.
> 
> All jobs were successfully interrupted part way through. How do the restart files look to you?
> 
> (use ls -lath to see log files from today near the top) **Log files:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/ **Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens/ **DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

Thanks, @AliciaBentley-NOAA! I didn't see the restart files for NAEFS grid2grid and precip jobs. I think NAEFS precip job didn't kick off (there is no log file). These two jobs do not need restart. You may try again if you'd like. Restart files for the rest of the jobs look good. Please proceed the restart runs.

GwenChen-NOAA committed:
Merge branch 'NOAA-EMC:develop' into grid2obs_restart

GwenChen-NOAA commented:
> @GwenChen-NOAA In addition to letting me know when I can run the remainder of the test, please sync your fork. A couple of PRs were merged this morning. Thanks!

Done.

AliciaBentley-NOAA commented:
@GwenChen-NOAA I think NAEFS grid2grid did run, but I agree that NAEFS precip didn't! Since it only runs for 4 minutes, I stopped it at 2 minutes. 

/lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/ 
-rw-r--r-- 1 alicia.bentley vpppg 3.3M Apr  4 15:56 jevs_global_ens_naefs_atmos_precip_stats.o28630842
-rw-r--r-- 1 alicia.bentley vpppg 3.1M Apr  4 12:41 jevs_global_ens_naefs_atmos_grid2grid_stats.o28607952

AliciaBentley-NOAA commented:
@GwenChen-NOAA @PerryShafran-NOAA After rerunning, there are now NAEFS precip restart files. There are still no restart files in NAEFS grid2grid (even after rerunning with a nearly full runtime) and I can't find anything in the NAEFS grid2grid log file where anything is being copied from stmp to COMOUT. I can find this in other log files. Am I missing something?

GwenChen-NOAA commented:
> @GwenChen-NOAA @PerryShafran-NOAA After rerunning, there are now NAEFS precip restart files. There are still no restart files in NAEFS grid2grid (even after rerunning with a nearly full runtime) and I can't find anything in the NAEFS grid2grid log file where anything is being copied from stmp to COMOUT. I can find this in other log files. Am I missing something?

Hmm, this needs a bit digging. I have restart files for NAEFS grid2grid in my test run. See /lfs/h2/emc/vpppg/noscrub/Lichuan.Chen/evs/v2.0/stats/global_ens/atmos.20250401/naefs/grid2grid

PerryShafran-NOAA commented:
> @GwenChen-NOAA @PerryShafran-NOAA After rerunning, there are now NAEFS precip restart files. There are still no restart files in NAEFS grid2grid (even after rerunning with a nearly full runtime) and I can't find anything in the NAEFS grid2grid log file where anything is being copied from stmp to COMOUT. I can find this in other log files. Am I missing something?

I think I'd suggest clearing the NAEFS grid2grid directory completely and running a full uninterrupted job, just to get a baseline and see how well this works.  

Also 6 minutes may not be long enough to get past any of the introductory processes in the job and get to any runs of metplus to get any stats files.  Some jobs have a lot to do to get to the point where they can start data processing.  

I *would* be concerned about these lines:

```
 gefsmbr read iret=          99
 cmcembr read iret=          99
 gefsmbr read iret=          99
 cmcembr read iret=          99
 gefsmbr read iret=          99
 cmcembr read iret=          99
 gefsmbr read iret=          99
 cmcembr read iret=          99
 gefsmbr read iret=          99
 cmcembr read iret=          99
```

A lot of iret=99 is concerning because that means that something is missing.  But perhaps the best way to go is to generate a single clean uninterrupted run and then go from there.

AliciaBentley-NOAA commented:
@PerryShafran-NOAA An example of the uninterrupted runs should be in the /naefs.20250401.saved directory in the same folder. Did the run not complete there either? 

CC @GwenChen-NOAA

PerryShafran-NOAA commented:
Oh, I didn't look.  Yeah, that looked fine.

What I might suggest is stopping the job at around 15 or 20 minutes.  It looks like the full run was a 47-minute job, and 6 minutes may not be enough to generate any data.  At least this is what it was appearing to me; it was still trying to pull in the prep data when the job was killed.  To give a true test, set the interrupt job at 15 or 20 minutes and see if you have any data in COMOUT.

AliciaBentley-NOAA commented:
@PerryShafran-NOAA jevs_global_ens_naefs_atmos_grid2grid_stats.sh ran for 47 minutes?? @GwenChen-NOAA told me the full length of jevs_global_ens_naefs_atmos_grid2grid_stats.sh was 0:08, which is why I stopped it at 00:06.

PerryShafran-NOAA commented:
> @PerryShafran-NOAA jevs_global_ens_naefs_atmos_grid2grid_stats.sh ran for 47 minutes?? @GwenChen-NOAA told me the full length of jevs_global_ens_naefs_atmos_grid2grid_stats.sh was 0:08, which is why I stopped it at 00:06.

Oh sorry, I'm looking at a grid2obs job.  Let me check the grid2grid job.

PerryShafran-NOAA commented:
@AliciaBentley-NOAA Your run of jevs_global_ens_naefs_atmos_grid2grid_stats.sh ran for 27 minutes.  This is the same amount of time it runs in emc.vpppg.

PerryShafran-NOAA commented:
I'm wondering if that 8 minutes is when Gwen had stopped her job for restart purposes, not the full runtime.  Some clarification is needed.  

@GwenChen-NOAA

AliciaBentley-NOAA commented:
@PerryShafran-NOAA Thanks! Perhaps a 00:06 wallclock time on a 27-minute job is not enough time. I thought it was an 8-minute job! I will let the job run for 15 minutes after the meeting that I'm currently in and we can see what happens. 

CC @GwenChen-NOAA

AliciaBentley-NOAA commented:
@GwenChen-NOAA @PerryShafran-NOAA I have kicked of the NAEFS grid2grid job again with a longer runtime (15 minutes) that will still kill it before it's full run time (27 minutes).

### jevs_global_ens_naefs_atmos_grid2grid_stats.sh
**Log file:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_naefs_atmos_grid2grid_stats.o28650102
**Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens/ 
**DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/jevs_global_ens_naefs_atmos_grid2grid_stats.28650102.dbqs01

GwenChen-NOAA commented:
> I'm wondering if that 8 minutes is when Gwen had stopped her job for restart purposes, not the full runtime. Some clarification is needed.
> 
> @GwenChen-NOAA

@PerryShafran-NOAA is right. NAEFS grid2grid job runs for 27+ min. I might mistype the runtime. It should be 28 min. Apologize for the typo.

AliciaBentley-NOAA commented:
Thanks, @GwenChen-NOAA ! That totally happens. In that case, I think the job is really close to hitting its allotted wallclock time instead of being much less than it. Here is the updated list of dev driver wallclock times to adjust at the end:

jevs_global_ens_cmce_atmos_grid2grid_stats.sh (runtime 0:28, requests 1:30, decrease wallclock requested)
jevs_global_ens_cmce_atmos_precip_stats.sh (runtime 0:04, requests 0:30, decrease wallclock requested)
jevs_global_ens_ecme_atmos_grid2obs_stats.sh (runtime 1:53, requests 2:00, **increase** wallclock requested)
jevs_global_ens_gefs_atmos_precip_stats.sh (runtime 0:06, requests 0:30, decrease wallclock requested)
jevs_global_ens_naefs_atmos_grid2grid_stats.sh (runtime 0:28, requests 0:30, **increase** wallclock requested)
jevs_global_ens_naefs_atmos_precip_stats.sh (runtime 0:04, requests 0:30, decrease wallclock requested)

AliciaBentley-NOAA commented:
@GwenChen-NOAA @PerryShafran-NOAA Running for 15 minutes produced some restart files! Can you please take a look before I kick everything off to try to complete these runs?

**Log file:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_naefs_atmos_grid2grid_stats.o28650102
**Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens/
**DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/jevs_global_ens_naefs_atmos_grid2grid_stats.28650102.dbqs01

GwenChen-NOAA commented:
> @GwenChen-NOAA @PerryShafran-NOAA Running for 15 minutes produced some restart files! Can you please take a look before I kick everything off to try to complete these runs?
> 
> **Log file:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_naefs_atmos_grid2grid_stats.o28650102 **Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens/ **DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/jevs_global_ens_naefs_atmos_grid2grid_stats.28650102.dbqs01

Looks good. Please proceed.

PerryShafran-NOAA commented:
@AliciaBentley-NOAA @GwenChen-NOAA For my part, the naefs grid2grid interrupt job looks good.  We have all the .completed files that exist in the working directory, and the stats files are located in the restart directory on COMOUT.  Looks like data is written to stmp and then copied over to COMOUT as it should.  Only errors that exist are ones associated with the job termination, so that's OK.  I think I'm satisfied that things are looking OK.

AliciaBentley-NOAA commented:
All jobs have been resubmitted with longer lead times to test the restart capability! 

**Log file:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/
**Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens/
**DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/

Job numbers are:
```
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_cmce_atmos_grid2grid_stats.sh 
28656183.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_cmce_atmos_grid2obs_stats.sh 
28656185.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_cmce_atmos_precip_stats.sh 
28656186.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_ecme_atmos_grid2grid_stats.sh 
28656191.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_ecme_atmos_grid2obs_stats.sh 
28656192.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_ecme_atmos_precip_stats.sh 
28656194.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_gefs_atmos_grid2grid_stats.sh 
28656227.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_gefs_atmos_grid2obs_stats.sh 
28656238.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_gefs_atmos_precip_stats.sh 
28656239.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_naefs_atmos_grid2grid_stats.sh 
28656256.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_naefs_atmos_grid2obs_stats.sh 
28656269.dbqs01
[alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_naefs_atmos_precip_stats.sh 
28656279.dbqs01
```

GwenChen-NOAA committed:
Adjust walltime and memory

GwenChen-NOAA committed:
Merge branch 'NOAA-EMC:develop' into grid2obs_restart

GwenChen-NOAA committed:
Merge branch 'grid2obs_restart' of https://github.com/GwenChen-NOAA/EVS into grid2obs_restart

GwenChen-NOAA commented:
> Thanks, @GwenChen-NOAA ! That totally happens. In that case, I think the job is really close to hitting its allotted wallclock time instead of being much less than it. Here is the updated list of dev driver wallclock times to adjust at the end:
> 
> jevs_global_ens_cmce_atmos_grid2grid_stats.sh (runtime 0:28, requests 1:30, decrease wallclock requested) jevs_global_ens_cmce_atmos_precip_stats.sh (runtime 0:04, requests 0:30, decrease wallclock requested) jevs_global_ens_ecme_atmos_grid2obs_stats.sh (runtime 1:53, requests 2:00, **increase** wallclock requested) jevs_global_ens_gefs_atmos_precip_stats.sh (runtime 0:06, requests 0:30, decrease wallclock requested) jevs_global_ens_naefs_atmos_grid2grid_stats.sh (runtime 0:28, requests 0:30, **increase** wallclock requested) jevs_global_ens_naefs_atmos_precip_stats.sh (runtime 0:04, requests 0:30, decrease wallclock requested)

@AliciaBentley-NOAA, I adjusted walltime and memory in the driver and ecf scripts based on the [EVS Best Practices document](https://docs.google.com/document/d/1Tve8cxESi-uaYcELHWk2Z-imNfqpHq3i2LKyRkZuEEY/edit?tab=t.0#heading=h.ey3fecfgzjs7), according to the numbers from the latest runs. I also made the numbers consistent in the driver and ecf scripts for the same job. Please take a look.

GwenChen-NOAA commented:
> All jobs have been resubmitted with longer lead times to test the restart capability!
> 
> **Log file:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/ **Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens/ **DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/
> 
> Job numbers are:
> 
> ```
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_cmce_atmos_grid2grid_stats.sh 
> 28656183.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_cmce_atmos_grid2obs_stats.sh 
> 28656185.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_cmce_atmos_precip_stats.sh 
> 28656186.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_ecme_atmos_grid2grid_stats.sh 
> 28656191.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_ecme_atmos_grid2obs_stats.sh 
> 28656192.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_ecme_atmos_precip_stats.sh 
> 28656194.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_gefs_atmos_grid2grid_stats.sh 
> 28656227.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_gefs_atmos_grid2obs_stats.sh 
> 28656238.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_gefs_atmos_precip_stats.sh 
> 28656239.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_naefs_atmos_grid2grid_stats.sh 
> 28656256.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_naefs_atmos_grid2obs_stats.sh 
> 28656269.dbqs01
> [alicia.bentley@dlogin03 global_ens] qsub jevs_global_ens_naefs_atmos_precip_stats.sh 
> 28656279.dbqs01
> ```

@AliciaBentley-NOAA, all restart jobs run successfully except NAEFS grid2grid job, which was killed due to exceed walltime. I have increased the walltime for this job from 30 min to 40 min in the last commit. Please try again.

AliciaBentley-NOAA commented:
@GwenChen-NOAA Thanks! I pulled in your 40-minute wallclock time and reran. 

### jevs_global_ens_naefs_atmos_grid2grid_stats.sh
**Log file:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/pr694test/EVS/dev/drivers/scripts/stats/global_ens/jevs_global_ens_naefs_atmos_grid2grid_stats.o28700535
**Output:** /lfs/h2/emc/vpppg/noscrub/alicia.bentley/evs/v2.0/stats/global_ens/
**DATA:** /lfs/h2/emc/stmp/alicia.bentley/evs_test/prod/tmp/jevs_global_ens_naefs_atmos_grid2grid_stats.28700535.dbqs01

AliciaBentley-NOAA commented:
@GwenChen-NOAA @PerryShafran-NOAA Please let me know if the rerun of `jevs_global_ens_naefs_atmos_grid2grid_stats.sh` that I did over the weekend looks good when you get the chance. Thanks!

PerryShafran-NOAA commented:
@AliciaBentley-NOAA It should be noted that the working directory in stmp is gone because of the retention period of stmp.  But it's good to know that the scrubber is working!  

The job ran for 10 minutes.  Correct me if I'm wrong, but this is a restart job, not the full job, yes?  If so, then this job time makes sense.  

The output stats file is the exact same size as the full-run final stats file.  No errors or warnings.

GwenChen-NOAA commented:
> @GwenChen-NOAA @PerryShafran-NOAA Please let me know if the rerun of `jevs_global_ens_naefs_atmos_grid2grid_stats.sh` that I did over the weekend looks good when you get the chance. Thanks!

@AliciaBentley-NOAA, same as @PerryShafran-NOAA, everything looks good to me.

AliciaBentley-NOAA reviewed (approved):
I have looked over the scripts changed and requested some wallclock time updates in the dev drivers. I have tested the scripts and reviewed the output. Any issues were commented on and subsequently fixed. I approve this restart bugzilla PR to be merged! 

CC @GwenChen-NOAA @PerryShafran-NOAA

PerryShafran-NOAA reviewed (approved):
Changes look good, code runs as expected.  I approve for merge.

AliciaBentley-NOAA commented:
@GwenChen-NOAA Can you please check that your fork is synced? I will merge you PR after that! Thanks.

GwenChen-NOAA commented:
> @GwenChen-NOAA Can you please check that your fork is synced? I will merge you PR after that! Thanks.

It is synced.
